### PR TITLE
[RDY] Enable -Wconversion for ex_cmds2.c

### PR DESCRIPTION
**Pull request by phanimahesh**
Ref #567.

This is my first take at enabling -Wconversion for ex_cmds2.c. It's been really long since I worked on C, so I'm not fully confident this is it. Comments welcome.

Where I thought necessary, I have added a bunch of assertions. Please point out if I missed any or added a little too many.

### CODE REVIEW

File: src/nvim/ex_cmds2.c [hunk 1]
```diff
@@ -1942,10 +1950,12 @@ void ex_listdo(exarg_T *eap)
         ex_cc(eap);
 
         buf = curbuf;
-        i = eap->line1 - 1;
+        assert(eap->line1 <= INT_MAX);
+        i = (int)eap->line1 - 1;
         if (eap->addr_count <= 0) {
           // Default to all quickfix/location list entries.
-          eap->line2 = qf_size;
+          assert(qf_size <= LONG_MAX);
```
Comment (phanimahesh) [inline src/nvim/ex_cmds2.c]: `linenr_T` is a type alias to `long`, and `qf_size` is `unsigned long` so I added this assertion. However it is brittle, if `linenr_T` changes, this will have to be updated. Is there a better way to do this?

**Comment by phanimahesh** [comment]
gcc is shouting at a couple of things clang isn't, failing the check. They look fine to me, though.

File: src/nvim/ex_cmds2.c (hunk 1 above)
Comment (bfredl) [inline src/nvim/ex_cmds2.c]: the correct bound is `< MAXLNUM`

File: src/nvim/ex_cmds2.c [hunk 2]
```diff
@@ -1847,19 +1852,21 @@ void ex_argdelete(exarg_T *eap)
     /* ":1,4argdel": Delete all arguments in the range. */
     if (eap->line2 > ARGCOUNT)
       eap->line2 = ARGCOUNT;
-    int n = eap->line2 - eap->line1 + 1;
+    long n = eap->line2 - eap->line1 + 1;
```
Comment (phanimahesh) [inline src/nvim/ex_cmds2.c]: should this be changed instead to `linenr_T` ?

File: src/nvim/ex_cmds2.c [hunk 3]
```diff
@@ -1847,19 +1852,21 @@ void ex_argdelete(exarg_T *eap)
     /* ":1,4argdel": Delete all arguments in the range. */
     if (eap->line2 > ARGCOUNT)
       eap->line2 = ARGCOUNT;
-    int n = eap->line2 - eap->line1 + 1;
```
Comment (bfredl) [inline src/nvim/ex_cmds2.c]: it would be more consistent to cast to `int` here, it should remove the warnings.

File: src/nvim/ex_cmds2.c [hunk 4]
```diff
@@ -1847,19 +1852,21 @@ void ex_argdelete(exarg_T *eap)
...
       EMSG(_(e_invarg));
     else {
-      for (int i = eap->line1; i <= eap->line2; ++i)
+      assert(eap->line1 <= INT_MAX);
+      assert(eap->line2 <= INT_MAX);
+      for (int i = (int)eap->line1; i <= eap->line2; ++i)
         xfree(ARGLIST[i - 1].ae_fname);
       memmove(ARGLIST + eap->line1 - 1, ARGLIST + eap->line2,
-          (size_t)((ARGCOUNT - eap->line2) * sizeof(aentry_T)));
+          (size_t)((unsigned long)(ARGCOUNT - eap->line2) * sizeof(aentry_T)));
```
Comment (bfredl) [inline src/nvim/ex_cmds2.c]: why not cast to `(size_t)` directly?

File: src/nvim/ex_cmds2.c [hunk 5]
```diff
@@ -1847,19 +1852,21 @@ void ex_argdelete(exarg_T *eap)
     /* ":1,4argdel": Delete all arguments in the range. */
     if (eap->line2 > ARGCOUNT)
       eap->line2 = ARGCOUNT;
-    int n = eap->line2 - eap->line1 + 1;
+    long n = eap->line2 - eap->line1 + 1;
     if (*eap->arg != NUL || n <= 0)
       EMSG(_(e_invarg));
     else {
-      for (int i = eap->line1; i <= eap->line2; ++i)
+      assert(eap->line1 <= INT_MAX);
```
Comment (phanimahesh) [inline src/nvim/ex_cmds2.c]: `w_arg_index` is int anyway, so I decided to cast linenumbers to int. I now have a bad feeling about this. Should we change loop index `i` to long instead, and cast to int only on 1869, since 1868 already guarantees it is safe?

File: src/nvim/ex_cmds2.c (hunk 4 above)
Comment (phanimahesh) [inline src/nvim/ex_cmds2.c]: My bad. Done.

File: src/nvim/ex_cmds2.c (hunk 3 above)
Comment (phanimahesh) [inline src/nvim/ex_cmds2.c]: But what is the guarantee that it will be an int? I can open a gigantic file and select all of it for deletion, it is a crazy use case, but shouldn't this be long, or more precisely `linenr_T` instead?

File: src/nvim/ex_cmds2.c (hunk 2 above)
Comment (bfredl) [inline src/nvim/ex_cmds2.c]: you could, but then you need to add the casts `(int)n` below. (Though the growarray `ga_len` should really be `size_t`, but that would likely affect a lot of stuff so let's keep it `int` for this PR)

File: src/nvim/ex_cmds2.c (hunk 3 above)
Comment (bfredl) [inline src/nvim/ex_cmds2.c]: For now linenumber differences are guaranteed to fit in an `int`. But you're right it's probably cleaner to keep it `linenr_T` as long as reasonable and then cast to an `int` when necessary.

File: src/nvim/ex_cmds2.c (hunk 5 above)
Comment (bfredl) [inline src/nvim/ex_cmds2.c]: why not make it `linenr_T i` ? Also these asserts are unneccesary.

Comment (phanimahesh) [inline src/nvim/ex_cmds2.c]: Just realized that while rereading this, done and pushed few minutes ago.

File: src/nvim/ex_cmds2.c [hunk 6]
```diff
@@ -3403,16 +3413,20 @@ static void script_host_execute_file(char *name, exarg_T *eap)
   // filename
   list_append_string(args, buffer, -1);
   // current range
-  list_append_number(args, eap->line1);
-  list_append_number(args, eap->line2);
+  assert(eap->line1 <= INT_MAX);
+  assert(eap->line2 <= INT_MAX);
+  list_append_number(args, (int)eap->line1);
+  list_append_number(args, (int)eap->line2);
```
Comment (phanimahesh) [inline src/nvim/ex_cmds2.c]: Is this right? Feels bad casting linenumbers to int unconditionally.

Comment (bfredl) [inline src/nvim/ex_cmds2.c]: It is fine, you can remove the `assert`s.

File: src/nvim/ex_cmds2.c [hunk 7]
```diff
@@ -72,7 +72,7 @@ typedef struct scriptitem_S {
   proftime_T sn_prl_start;      /* start time for current line */
   proftime_T sn_prl_children;    /* time spent in children for this line */
   proftime_T sn_prl_wait;       /* wait start time for current line */
-  int sn_prl_idx;               /* index of line being timed; -1 if none */
+  long sn_prl_idx;               /* index of line being timed; -1 if none */
```
Comment (cacplate) [inline src/nvim/ex_cmds2.c]: Can you do without adding more `long`'s? It's not permitted from style guide.

File: src/nvim/ex_cmds2.c [hunk 8]
```diff
@@ -1255,7 +1257,7 @@ check_changed_any (
   int save;
   int i;
   int bufnum = 0;
-  int bufcount = 0;
+  unsigned int bufcount = 0;
```
Comment (cacplate) [inline src/nvim/ex_cmds2.c]: better be size_t I think. More consistent with the rest of the codebase.

File: src/nvim/ex_cmds2.c [hunk 9]
```diff
@@ -57,23 +57,23 @@ typedef struct scriptitem_S {
...
+  proftime_T sn_pr_total;       // time spent in script + children
+  proftime_T sn_pr_self;        // time spent in script itself
+  proftime_T sn_pr_start;       // time at script start
+  proftime_T sn_pr_children;    // time in children after script start
+  // profiling the script per line
+  garray_T sn_prl_ga;           // things stored for every line
+  proftime_T sn_prl_start;      // start time for current line
+  proftime_T sn_prl_children;    // time spent in children for this line
+  proftime_T sn_prl_wait;       // wait start time for current line
+  long sn_prl_idx;               // index of line being timed; -1 if none
```
Comment (phanimahesh) [inline src/nvim/ex_cmds2.c]: @cacplate Should I cast things to int instead? Since this is index of a line, probably it should be linenr_T?

File: src/nvim/ex_cmds2.c (hunk 8 above)
Comment (phanimahesh) [inline src/nvim/ex_cmds2.c]: Thanks. Done.

File: src/nvim/ex_cmds2.c (hunk 6 above)
Comment (phanimahesh) [inline src/nvim/ex_cmds2.c]: Can you please explain why is it safe to remove asserts?

Comment (bfredl) [inline src/nvim/ex_cmds2.c]: Because a linenumber is guaranteed to fit in an `int` (yes, one day we might want to lift the restriction and allow more than 2G lines, however this would be a long-term effort and I don't think cluttering our codebase with asserts would help towards that goal)

Comment (phanimahesh) [inline src/nvim/ex_cmds2.c]: Makes sense, but if that is the case, what is the rationale behind having `linenr_T` as an alias for `long` instead of an `int`?

If we have a casting where code does not guarantee it is safe, I prefer to add an assertion. line numbers in any sane use case will fit in an int, but technically linenr_T is a long, so..

Apologies for being persistent, I'm trying to understand the rationale here.

Comment (bfredl) [inline src/nvim/ex_cmds2.c]: Likely just because vim does it, and noone has seriously suggested changing it :) Note as vim (and neovim) compiles works just fine on ILP32 or IL32P64 (64-bit windows) it cannot be that vim relies on `linenr_T > INT_MAX` as sentinel values or say temporary overflow in the midst of a calculation.
I don't like adding an assert for every technicallity, it makes sense on some casts but here it just adds noise. Rather we should continue to slowly but steadly refactor to use better and more consistent int types, to reduce the number of casts needed. (It might be worthwhile to try change `linenr_T` to say `int32_t` in the short run but it could generate warnings and follow-on changes, so it might not be an obvious change, but could be worth trying.)

File: src/nvim/ex_cmds2.c [hunk 10]
```diff
@@ -57,23 +57,23 @@ typedef struct scriptitem_S {
...
-  proftime_T sn_pr_start;       /* time at script start */
-  proftime_T sn_pr_children;    /* time in children after script start */
-  /* profiling the script per line */
-  garray_T sn_prl_ga;           /* things stored for every line */
-  proftime_T sn_prl_start;      /* start time for current line */
-  proftime_T sn_prl_children;    /* time spent in children for this line */
-  proftime_T sn_prl_wait;       /* wait start time for current line */
-  int sn_prl_idx;               /* index of line being timed; -1 if none */
-  int sn_prl_execed;            /* line being timed was executed */
+  bool sn_prof_on;              // true when script is/was profiled
```
Comment (cacplate) [inline src/nvim/ex_cmds2.c]: you could use `///<` style comments for doc.

File: src/nvim/ex_cmds2.c [hunk 11]
```diff
@@ -417,8 +419,8 @@ static garray_T prof_ga = {0, 0, sizeof(struct debuggy), 4, NULL};
  * is allocated.
  * Returns FAIL for failure.
  */
-static int 
-dbg_parsearg (
+static int
+dbg_parsearg(
```
Comment (cacplate) [inline src/nvim/ex_cmds2.c]: type, name and arguments on same line + doxygen docstring if possible

File: src/nvim/ex_cmds2.c [hunk 12]
```diff
@@ -551,7 +553,7 @@ void ex_debuggreedy(exarg_T *eap)
 void ex_breakdel(exarg_T *eap)
 {
   struct debuggy *bp, *bpi;
-  int nr;
+  long nr;
```
Comment (cacplate) [inline src/nvim/ex_cmds2.c]: same, no `long` if possible.

File: src/nvim/ex_cmds2.c [hunk 13]
```diff
@@ -641,11 +645,11 @@ void ex_breaklist(exarg_T *eap)
  * Find a breakpoint for a function or sourced file.
  * Returns line number at which to break; zero when no matching breakpoint.
  */
-linenr_T 
-dbg_find_breakpoint (
-    int file,                   /* TRUE for a file, FALSE for a function */
-    char_u *fname,         /* file or function name */
-    linenr_T after             /* after this line number */
+linenr_T
+dbg_find_breakpoint(
```
Comment (cacplate) [inline src/nvim/ex_cmds2.c]: same as the other comment.

File: src/nvim/ex_cmds2.c [hunk 14]
```diff
@@ -641,11 +645,11 @@ void ex_breaklist(exarg_T *eap)
...
+    int file,                   // TRUE for a file, FALSE for a function
+    char_u *fname,              // file or function name
+    linenr_T after              // after this line number
```
Comment (cacplate) [inline src/nvim/ex_cmds2.c]: Please refrain from making unrelated changes. This has nothing to do with Wconversion so you better make another PR that refactor function signatures.

Comment (phanimahesh) [inline src/nvim/ex_cmds2.c]: This was to silent clint, it was warning that old style-comments are not allowed, and function name should not be followed by a space before parenthesis. Travis build was failing otherwise.

File: src/nvim/ex_cmds2.c (hunk 6 above)
Comment (phanimahesh) [inline src/nvim/ex_cmds2.c]: Makes sense. I'll nuke the assertions.

File: src/nvim/ex_cmds2.c (hunk 12 above)
Comment (phanimahesh) [inline src/nvim/ex_cmds2.c]: Possible if I change `atol((char *)eap->arg)` below this to `atoi((char *)eap->arg)`. Has same effect as casting the result to int. Done.

File: src/nvim/ex_cmds2.c (hunk 14 above)
Comment (bfredl) [inline src/nvim/ex_cmds2.c]: I think changing signature doc is fine if you makes changes to signature anyways (or obviously if required by the linter as in this case), but why not change to the correct doxygen style directly? (if you are not familiar look i e in `charset.c` or `diff.c` which already are refactored)

File: src/nvim/ex_cmds2.c (hunk 6 above)
Comment (KillTheMule) [inline src/nvim/ex_cmds2.c]: I've run into those issues in https://github.com/neovim/neovim/pull/4076, where I made a suggestion to remedy, but the project has not come forth with a decision.

Comment (bfredl) [inline src/nvim/ex_cmds2.c]: your points in #4076 seem reasonable but maybe rename or break out the `linenr_T` part to a separate issue/PR because changing `linenr_T` seems like a bigger change/concern than fixing conversion in `diff.c`

**Comment by phanimahesh** [comment]
I have rebased this on top of master, squashed and reordered commits. Pushing in a few minutes.

There is one commit for code changes due to -Wconversion, followed by a sequence of commits each dealing with style fixes to please clint, transition to doxygen style docs, and to adhere to the style guide.

**Comment by phanimahesh** [comment]
I am trying to clean up the file completely. If the cleanup is desired in a different PR, I can just cherrypick commits from here since I'm making one commit per function/struct/doc block. Please let me know.

**Comment by cacplate** [comment]
Better make a different PR but the linting warnings from the Wconversion changes belong in this PR. The rest can do another.

**Comment by phanimahesh** [comment]
The trouble is, seemingly irrelevant changes are triggering clint warnings, possibly because context changed and no longer match the whitelist in https://raw.githubusercontent.com/neovim/doc/gh-pages/reports/clint/errors.json.

I'll try to find the smallest subset of edits to appease clint and move everything else to a different pr by tomorrow.

**Comment by cacplate** [comment]
Whenever you make a change this triggers Clint warnings for adjacent lines that trigger warnings for their own adjacent lines once they are fixed. In the end a whole block of variables with comments need to be changed but that's OK.

**Comment by phanimahesh** [comment]
This pr now contains only the minimal set of style changes required to satisfy clint. I'll open a new PR for rest of the clint changes. This can go `[RDY]` after the ci builds report success.

The struct docs technically can be split into multiple commits and only one block of them is needed to satisfy clint, but I let the others stay to keep things consistent, and they don't introduce much noise either.

@cacplate, @bfredl: I have addressed all comments that have been made, please review.

File: src/nvim/ex_cmds2.c [hunk 15]
```diff
@@ -272,7 +272,8 @@ void do_debug(char_u *cmd)
 
       xfree(cmdline);
     }
-    lines_left = Rows - 1;
+    assert(Rows - 1 <= INT_MAX);
```
Comment (bfredl) [inline src/nvim/ex_cmds2.c]: not needed (`Rows` is limited in `screen.c:limit_screen_size`)

File: src/nvim/ex_cmds2.c [hunk 16]
```diff
@@ -1702,10 +1707,12 @@ void ex_argument(exarg_T *eap)
 {
   int i;
 
-  if (eap->addr_count > 0)
-    i = eap->line2 - 1;
-  else
+  if (eap->addr_count > 0) {
+    assert(eap->line2 <= INT_MAX);
```
Comment (bfredl) [inline src/nvim/ex_cmds2.c]: not needed

File: src/nvim/ex_cmds2.c [hunk 17]
```diff
@@ -1909,7 +1919,8 @@ void ex_listdo(exarg_T *eap)
         }
         break;
       case CMD_argdo:
-        i = eap->line1 - 1;
+        assert(eap->line1 <= INT_MAX);
```
Comment (bfredl) [inline src/nvim/ex_cmds2.c]: ditto

File: src/nvim/ex_cmds2.c [hunk 18]
```diff
@@ -93,18 +93,18 @@ typedef struct sn_prl_S {
  * sourcing can be done recursively.
  */
 struct source_cookie {
-  FILE        *fp;              /* opened file for sourcing */
-  char_u      *nextline;        /* if not NULL: line that was read ahead */
-  int finished;                 /* ":finish" used */
+  FILE        *fp;              ///< opened file for sourcing
```
Comment (cacplate) [inline src/nvim/ex_cmds2.c]: better remove extra spaces here too.

File: src/nvim/ex_cmds2.c [hunk 19]
```diff
@@ -93,18 +93,18 @@ typedef struct sn_prl_S {
...
+  int fileformat;               ///< EOL_UNKNOWN, EOL_UNIX or EOL_DOS
+  int error;                    ///< TRUE if LF found after CR-LF
 #endif
-  linenr_T breakpoint;          /* next line with breakpoint or zero */
-  char_u      *fname;           /* name of sourced file */
-  int dbg_tick;                 /* debug_tick when breakpoint was set */
-  int level;                    /* top nesting level of sourced file */
-  vimconv_T conv;               /* type of conversion */
+  linenr_T breakpoint;          ///< next line with breakpoint or zero
+  char_u      *fname;           ///< name of sourced file
```
Comment (cacplate) [inline src/nvim/ex_cmds2.c]: ditto

File: src/nvim/ex_cmds2.c [hunk 20]
```diff
@@ -392,12 +392,12 @@ int dbg_check_skipped(exarg_T *eap)
  * This is a grow-array of structs.
  */
 struct debuggy {
-  int dbg_nr;                   /* breakpoint number */
```
Comment (cacplate) [inline src/nvim/ex_cmds2.c]: This change and more before it change nothing except comment style, that's the kind of thing that can go in another PR. Maybe I'm being too picky. @bfredl what do you think, can we just leave it?

File: src/nvim/ex_cmds2.c [hunk 21]
```diff
@@ -392,12 +392,12 @@ int dbg_check_skipped(exarg_T *eap)
...
-  int dbg_type;                 /* DBG_FUNC or DBG_FILE */
-  char_u      *dbg_name;        /* function or file name */
-  regprog_T   *dbg_prog;        /* regexp program */
-  linenr_T dbg_lnum;            /* line number in function or file */
-  int dbg_forceit;              /* ! used */
+  int dbg_nr;                   ///< breakpoint number
+  int dbg_type;                 ///< DBG_FUNC or DBG_FILE
+  char_u      *dbg_name;        ///< function or file name
```
Comment (cacplate) [inline src/nvim/ex_cmds2.c]: more spaces removal

File: src/nvim/ex_cmds2.c (hunk 20 above)
Comment (bfredl) [inline src/nvim/ex_cmds2.c]: I think its fine, because this keeps it consistent with the struct that was needed to change.

Comment (phanimahesh) [inline src/nvim/ex_cmds2.c]: Yep, I thought changing just one struct's comment style makes things inconsistent, and changed all structs and enums in one go. Figured it's a relatively small additional noise but is okay because it keeps things consistent.

**Comment by phanimahesh** [comment]
@cacplate There are other places with extra spaces too but I'm leaving them as-is. Touching them is going to trigger a few more clint warnings, and I am already cleaning up all clint warnings in a separate branch.

**Comment by phanimahesh** [comment]
Any blockers for merging this?

File: src/nvim/ex_cmds2.c [hunk 22]
```diff
@@ -93,18 +93,18 @@ typedef struct sn_prl_S {
...
-  char_u      *nextline;        /* if not NULL: line that was read ahead */
-  int finished;                 /* ":finish" used */
+  FILE *fp;                     ///< opened file for sourcing
+  char_u *nextline;             ///< if not NULL: line that was read ahead
+  int finished;                 ///< ":finish" used
 #if defined(USE_CRNL)
-  int fileformat;               /* EOL_UNKNOWN, EOL_UNIX or EOL_DOS */
-  int error;                    /* TRUE if LF found after CR-LF */
+  int fileformat;               ///< EOL_UNKNOWN, EOL_UNIX or EOL_DOS
+  int error;                    ///< TRUE if LF found after CR-LF
```
Comment (bfredl) [inline src/nvim/ex_cmds2.c]: true

File: src/nvim/ex_cmds2.c [hunk 23]
```diff
@@ -1942,10 +1949,12 @@ void ex_listdo(exarg_T *eap)
         ex_cc(eap);
 
         buf = curbuf;
-        i = eap->line1 - 1;
+        assert(eap->line1 <= INT_MAX);
```
Comment (bfredl) [inline src/nvim/ex_cmds2.c]: remove this one as well.

**Comment by bfredl** [comment]
LGTM except two minor comments. Please squash the two style commits to one.

File: src/nvim/ex_cmds2.c (hunk 22 above)
Comment (phanimahesh) [inline src/nvim/ex_cmds2.c]: The type has not been converted to bool yet. Should I still change the comment?

Comment (phanimahesh) [inline src/nvim/ex_cmds2.c]: This and a few other fields can be changed to bool, since the struct is fully local to this file. That can go in with another PR also handling all clint warnings and updating docs, maybe?

I'm working on that partly, waiting for this to be merged so that I can do a rebase and start another PR.

Comment (bfredl) [inline src/nvim/ex_cmds2.c]: OK, leave it then :+1:

**Comment by phanimahesh** [comment]
Done. Also rebased on current master.

**Comment by bfredl** [comment]
merged, thanks!
